### PR TITLE
[performance](pipelineX) optimize pipelineX

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes

1. Use vectors for "get local state" instead of maps, and it doesn't require locking.
2. Clearly distinguish between node ID and operator ID.
3. Remove unnecessary locks from dependencies.
4. For pipeline tasks, directly obtain dependencies instead of using operators, and call "get local state."
5. Use unique pointers instead of shared pointers.

 

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

File: be/src/pipeline/pipeline_x/pipeline_x_task.h [hunk 1]
```diff
@@ -153,6 +155,9 @@ class PipelineXTask : public PipelineTask {
     OperatorXPtr _root;
     DataSinkOperatorXPtr _sink;
 
+    std::vector<Dependency*> _operatorsDependency;
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/pipeline_x_task.h:158]:
```suggestion
    std::vector<Dependency*> _read_dependencies;
```

File: be/src/pipeline/pipeline_x/pipeline_x_task.h [hunk 2]
```diff
@@ -153,6 +155,9 @@ class PipelineXTask : public PipelineTask {
...
+    WriteDependency* _sinkDependency;
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/pipeline_x_task.h:159]:
```suggestion
    WriteDependency* _write_dependencies;
```

File: be/src/pipeline/pipeline_x/pipeline_x_task.h [hunk 3]
```diff
@@ -141,6 +141,8 @@ class PipelineXTask : public PipelineTask {
         return _upstream_dependency[id];
     }
 
+    Status get_dependency_from_task();
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/pipeline_x_task.h:144]:
```suggestion
    Status extract_dependencies();
```

File: be/src/pipeline/pipeline_x/dependency.h [hunk 4]
```diff
@@ -88,15 +89,13 @@ class Dependency : public std::enable_shared_from_this<Dependency> {
...
         child->set_parent(weak_from_this());
     }
 
-    void remove_first_child() {
-        std::unique_lock<std::mutex> l(_lock);
-        _children.erase(_children.begin());
-    }
+    void remove_first_child() { _children.erase(_children.begin()); }
+
+    virtual bool is_not_fake() const { return true; }
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/dependency.h:98]: Maybe delete this function

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**